lib/json.js

```diff
@@ -0,0 +1,60 @@
+function formatNumber(num, precision = 6) {
+  if (typeof num !== "number") {
+    return String(num);
+  } else if (isNaN(num)) {
+    return `null`;
+  } else if (num === 0) {
+    return "0";
+  }
+
+  const absNum = Math.abs(num);
+  // Python general format uses scientific notation if exponent >= precision or <= -4
+  const exponent = Math.floor(Math.log10(absNum));
+  const useScientific = exponent >= precision || exponent < -4;
+
+  if (useScientific) {
+    // Use scientific notation
+    return num
+      .toExponential(precision - 1)
+      .replace(/\.?0+e/, "e") // Remove trailing zeros before 'e'
+      .replace(/\.$/, "") // Remove trailing decimal point
+      .replace(/e\+?/, "e") // Remove '+' sign and leading zeros in positive exponent
+      .replace(/e-0?/, "e-"); // Remove leading zeros in negative exponent
+  } else {
+    // Use fixed notation
+    // Calculate decimal places needed based on exponent and precision
+    const decimalPlaces = Math.max(0, precision - exponent - 1);
+    return num
+      .toFixed(decimalPlaces)
+      .replace(/\.?0+$/, "") // Remove trailing zeros
+      .replace(/\.$/, ""); // Remove trailing decimal point
+  }
+}
+
+/**
+ * Encodes a JavaScript data structure to JSON, reformatting all numbers encountered
+ * along the way in a consistent, predictable fashion.
+ * @param {*} data
+ * @returns String
+ */
+function serialize(data, indent = 2) {
+  const token = ":|--|:";
+  return JSON.stringify(
+    data,
+    (_, value) => {
+      if (typeof value === "number") {
+        return token + formatNumber(value) + token;
+      } else {
+        return value;
+      }
+    },
+    indent,
+  )
+    .replaceAll(`"${token}`, "")
+    .replaceAll(`${token}"`, "");
+}
+
+module.exports = {
+  formatNumber,
+  serialize,
+};
```

Comment:
Like in Python, wouldn't it be much shorter to use the `replace` argument of JSON.Stringify ?
Something like:
```
function formatJSON(obj, precision = 6) {
  return JSON.stringify(obj, (key, value) => {
    if (typeof value === "number") {
      return Number(value.toPrecision(precision));
    }
    return value;
  }, 0);
}
```
then
```
> formatJSON({"a": 1.23456789e-5, "b": [1.23456789e-5], "c": 1.23456789e-5})
'{"a":0.0000123457,"b":[0.0000123457],"c":0.0000123457}'
```

Reply:
ToPrecision doesn't implement the same behavior, it goes from decimal to scientific at 1e-6
<img width="1149" alt="Screenshot 2024-12-05 at 16 10 20" src="https://github.com/user-attachments/assets/14e5c9b4-42f4-477a-928f-33f98a5f8e3c">
<img width="768" alt="Screenshot 2024-12-05 at 16 12 28" src="https://github.com/user-attachments/assets/2ed2d7d2-55c9-44e6-925e-6899234c3f5c">
https://jsfiddle.net/9knf4j7x/